### PR TITLE
Added sniem recipe

### PR DESCRIPTION
**Pull request by SpringHan**
### Brief summary of what the package does

Sniem means Simple United Editing Method. It's a new editing method for emacs. All the motions and most of the operations are related to `last-point`, and there are some great functions like `sniem-object-catch`, `sniem-middle-keyboard-or-fn`, `sniem-mark-jump`, etc.

### Direct link to the package repository

https://github.com/SpringHan/sniem

### Your association with the package

I'm the maintainer.

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

Please confirm with `x`:

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [ ] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

### CODE REVIEW

File: recipes/sniem [hunk 1]
```diff
@@ -0,0 +1 @@
+(sniem :fetcher github :repo "SpringHan/sniem" :files ("*.el"))
```
Comment (riscy) [inline recipes/sniem:1]: Default recipe should suffice here:
```suggestion
(sniem :repo "SpringHan/sniem" :fetcher github)
```

**Review by riscy — COMMENTED** [review]
Thanks for this.  There's a lot of feedback below, but I'm happy to help clarify anything.  Multi-file packages can be tricky to get right.

One key item is that your `package-requires` line in sniem.el:
`;; Package-Requires: ((emacs) (s "20180406.808") (dash "20200524.1947"))`
should specify an Emacs version e.g. `(emacs "26.1")` (you will see lots of the package-lint feedback below is pressing you for an Emacs version).

### sniem-cheatsheet.el ###

byte-compile (using Emacs 27.1):
```
sniem-cheatsheet.el:250:13:Warning: reference to free variable
    `sniem-keyboard-layout'
In sniem-show-cheatsheet:
sniem-cheatsheet.el:261:8:Warning: sniem-show-cheatsheet being defined to take
    1 arg, but was previously called with 0
sniem-cheatsheet.el:279:4:Warning: `beginning-of-buffer' is for interactive
    use only; use `(goto-char (point-min))' instead.
In sniem-cheatsheet--replace:
sniem-cheatsheet.el:282:8:Warning: sniem-cheatsheet--replace being defined to
    take 1 arg, but was previously called with 0
sniem-cheatsheet.el:297:49:Warning: reference to free variable
    `sniem-leader-keymap'
In sniem-cheatsheet--command:
sniem-cheatsheet.el:317:20:Warning: reference to free variable
    `number-to-string'
In end of data:
sniem-cheatsheet.el:334:1:Warning: the following functions are not known to be
    defined: ->, -reduce-from, s-replace, ->>, s-truncate
```

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20210127.158):
```
10 issues found:
6:0: error: Package-Requires outside the main file have no effect.
8:3: warning: You should include standard keywords: see the variable `finder-known-keywords'.
251:7: error: You should depend on (emacs "24.3") if you need `user-error'.
252:30: error: You should depend on (emacs "24.1") if you need `pcase'.
265:5: error: You should depend on (emacs "24.3") if you need `setq-local'.
276:5: error: You should depend on (emacs "26.1") if you need `display-line-numbers-mode'.
278:5: error: You should depend on (emacs "24.3") if you need `setq-local'.
293:37: error: You should depend on (emacs "25.1") if you need `alist-get'.
308:19: error: You should depend on (emacs "25.1") if you need `alist-get'.
311:25: error: You should depend on (emacs "25.1") if you need `alist-get'.
```

### sniem-macro.el ###

byte-compile (using Emacs 27.1):
- No issues!

checkdoc (using version 0.6.2):
```
Warning (emacs): 
sniem-macro.el:34: Argument `name' should appear (as NAME) in the doc string
```

package-lint (using version 20210127.158):
```
3 issues found:
6:0: error: Package-Requires outside the main file have no effect.
8:3: warning: You should include standard keywords: see the variable `finder-known-keywords'.
46:10: error: You should depend on (emacs "24.3") if you need `setq-local'.
```

### sniem-mark-jump.el ###

byte-compile (using Emacs 27.1):
```
sniem-mark-jump.el:91:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-mark-jump.el:91:22:Warning: reference to free variable
    `sniem-mark-jump-next'
sniem-mark-jump.el:91:54:Warning: reference to free variable `type'
sniem-mark-jump.el:96:28:Warning: assignment to free variable `type'
sniem-mark-jump.el:98:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-mark-jump.el:98:22:Warning: reference to free variable
    `sniem-mark-jump-prev'
In end of data:
sniem-mark-jump.el:185:1:Warning: the following functions are not known to be
    defined: sniem-open-line-previous, sniem-object-catch-lisp-mode-p,
    sniem-change-mode, sniem-define-motion, &optional, sniem--mems
```

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20210127.158):
```
5 issues found:
6:0: error: Package-Requires outside the main file have no effect.
8:3: warning: You should include standard keywords: see the variable `finder-known-keywords'.
79:5: error: You should depend on (emacs "25.1") if you need `save-mark-and-excursion'.
154:7: error: You should depend on (emacs "24.3") if you need `setq-local'.
176:3: error: You should depend on (emacs "24.1") if you need `pcase'.
```

### sniem-object-catch.el ###

byte-compile (using Emacs 27.1):
```
sniem-object-catch.el:37:1:Warning: defcustom for
    `sniem-object-catch-global-symbol-alist' fails to specify type
sniem-object-catch.el:37:1:Warning: defcustom for
    `sniem-object-catch-global-symbol-alist' fails to specify type
sniem-object-catch.el:70:1:Warning: misplaced interactive spec:
    `(interactive)'
sniem-object-catch.el:70:22:Warning: reference to free variable
    `sniem-object-catch'
sniem-object-catch.el:70:52:Warning: reference to free variable `char'
sniem-object-catch.el:70:57:Warning: reference to free variable `parent'
In sniem-object-catch-parent-by-char:
sniem-object-catch.el:175:32:Warning: `message' called with 0 args to fill 1
    format field(s)
In prog3:
sniem-object-catch.el:296:11:Warning: macro `prog3' defined too late
sniem-object-catch.el:305:1:Warning: Lexical argument shadows the dynamic
    variable mode-name
In end of data:
sniem-object-catch.el:334:1:Warning: the following functions are not known to
    be defined: sniem-define-motion, &optional, sniem-object-catch,
    sniem-normal-set-key
```

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20210127.158):
```
14 issues found:
6:0: error: Package-Requires outside the main file have no effect.
8:3: warning: You should include standard keywords: see the variable `finder-known-keywords'.
67:0: error: "global-sniem-object-catch-status" doesn't start with package's prefix "sniem".
81:7: error: You should depend on (emacs "24.3") if you need `setq-local'.
91:5: error: You should depend on (emacs "25.1") if you need `save-mark-and-excursion'.
122:9: error: You should depend on (emacs "24.3") if you need `setq-local'.
136:11: error: You should depend on (emacs "24.3") if you need `setq-local'.
139:5: error: You should depend on (emacs "24.3") if you need `setq-local'.
188:3: error: You should depend on (emacs "24.3") if you need `setq-local'.
225:5: error: You should depend on (emacs "25.1") if you need `save-mark-and-excursion'.
248:3: error: You should depend on (emacs "25.1") if you need `save-mark-and-excursion'.
296:0: error: "prog3" doesn't start with package's prefix "sniem".
321:39: error: You should depend on (emacs "24.3") if you need `setq-local'.
323:39: error: You should depend on (emacs "24.3") if you need `setq-local'.
```

Other [lints](https://github.com/riscy/melpazoid):
- sniem-object-catch.el#L319: There is no need to quote lambdas (neither #' nor ')
- sniem-object-catch.el#L319: Loading a package should rarely add hooks

### sniem-operation.el ###

byte-compile (using Emacs 27.1):
```
sniem-operation.el:90:10:Warning: `beginning-of-buffer' is for interactive use
    only; use `(goto-char (point-min))' instead.
In sniem-center:
sniem-operation.el:99:33:Warning: reference to free variable
    `sniem-center-message'
In sniem-mark:
sniem-operation.el:107:33:Warning: reference to free variable
    `sniem-mark-message'
sniem-operation.el:113:18:Warning: assignment to free variable
    `sniem-mark-line'
sniem-operation.el:115:17:Warning: reference to free variable
    `sniem-last-point'
sniem-operation.el:116:12:Warning: reference to free variable
    `sniem-last-point-locked'
sniem-operation.el:122:43:Warning: reference to free variable
    `sniem-mark-line'
sniem-operation.el:123:67:Warning: assignment to free variable
    `sniem-mark-line'
In sniem-up/down-case:
sniem-operation.el:131:13:Warning: `delete-forward-char' is for interactive
    use only; use `delete-char' instead.
In sniem-replace-char:
sniem-operation.el:137:4:Warning: `delete-forward-char' is for interactive use
    only; use `delete-char' instead.
In sniem-delete:
sniem-operation.el:163:35:Warning: reference to free variable
    `sniem-delete-message'
sniem-operation.el:179:31:Warning: reference to free variable
    `sniem-last-point'
sniem-operation.el:180:16:Warning: reference to free variable
    `sniem-last-point-locked'
In sniem-change:
sniem-operation.el:205:35:Warning: reference to free variable
    `sniem-change-message'
In sniem-yank:
sniem-operation.el:226:35:Warning: reference to free variable
    `sniem-yank-message'
sniem-operation.el:233:26:Warning: reference to free variable
    `sniem-last-point'
sniem-operation.el:234:16:Warning: reference to free variable
    `sniem-last-point-locked'
In sniem-macro:
sniem-operation.el:337:35:Warning: reference to free variable
    `sniem-macro-message'
sniem-operation.el:342:15:Warning: reference to free variable
    `sniem-kmacro-range'
sniem-operation.el:362:25:Warning: assignment to free variable
    `sniem-kmacro-range'
sniem-operation.el:368:54:Warning: assignment to free variable
    `sniem-kmacro-mark-content'
sniem-operation.el:376:21:Warning: reference to free variable
    `sniem-kmacro-mark-content'
sniem-operation.el:377:55:Warning: assignment to free variable
    `sniem-kmacro-mark-content'
sniem-operation.el:379:1:Warning: Unused lexical variable `second-char'
sniem-operation.el:426:22:Warning: reference to free variable
    `sniem-beginning-of-line'
sniem-operation.el:430:22:Warning: reference to free variable
    `sniem-end-of-line'
sniem-operation.el:434:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-operation.el:434:22:Warning: reference to free variable
    `sniem-forward-char'
sniem-operation.el:434:52:Warning: reference to free variable `n'
sniem-operation.el:437:15:Warning: assignment to free variable `n'
sniem-operation.el:445:22:Warning: reference to free variable
    `sniem-5-forward-char'
sniem-operation.el:449:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-operation.el:449:22:Warning: reference to free variable
    `sniem-backward-char'
sniem-operation.el:460:22:Warning: reference to free variable
    `sniem-5-backward-char'
sniem-operation.el:464:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-operation.el:464:22:Warning: reference to free variable
    `sniem-prev-line'
sniem-operation.el:467:15:Warning: `previous-line' is for interactive use
    only; use `forward-line' with negative argument instead.
sniem-operation.el:473:22:Warning: reference to free variable
    `sniem-5-prev-line'
sniem-operation.el:477:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-operation.el:477:22:Warning: reference to free variable
    `sniem-next-line'
sniem-operation.el:480:15:Warning: `next-line' is for interactive use only;
    use `forward-line' instead.
sniem-operation.el:486:22:Warning: reference to free variable
    `sniem-5-next-line'
sniem-operation.el:490:22:Warning: reference to free variable
    `sniem-first-line'
sniem-operation.el:496:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-operation.el:496:22:Warning: reference to free variable
    `sniem-goto-line'
sniem-operation.el:506:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-operation.el:506:22:Warning: reference to free variable
    `sniem-scroll-up-command'
sniem-operation.el:513:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-operation.el:513:22:Warning: reference to free variable
    `sniem-scroll-down-command'
sniem-operation.el:520:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-operation.el:520:22:Warning: reference to free variable
    `sniem-find-forward'
sniem-operation.el:520:54:Warning: reference to free variable `c'
sniem-operation.el:520:56:Warning: reference to free variable `no-hint'
sniem-operation.el:536:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-operation.el:536:22:Warning: reference to free variable
    `sniem-find-backward'
sniem-operation.el:564:1:Warning: Unused lexical variable `point'
sniem-operation.el:564:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-operation.el:564:22:Warning: reference to free variable
    `sniem-next-word'
sniem-operation.el:564:59:Warning: reference to free variable `word'
sniem-operation.el:583:1:Warning: Unused lexical variable `point'
sniem-operation.el:583:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-operation.el:583:22:Warning: reference to free variable
    `sniem-prev-word'
sniem-operation.el:604:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-operation.el:604:22:Warning: reference to free variable
    `sniem-next-symbol'
sniem-operation.el:613:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-operation.el:613:22:Warning: reference to free variable
    `sniem-prev-symbol'
sniem-operation.el:622:22:Warning: reference to free variable
    `sniem-beg-of-mark'
sniem-operation.el:629:22:Warning: reference to free variable
    `sniem-end-of-mark'
sniem-operation.el:636:22:Warning: reference to free variable
    `sniem-goto-prev'
sniem-operation.el:642:22:Warning: reference to free variable
    `sniem-goto-next'
In sniem-goto-last-point:
sniem-operation.el:653:20:Warning: reference to free variable
    `sniem-last-goto-point'
sniem-operation.el:655:18:Warning: reference to free variable
    `sniem-last-point'
sniem-operation.el:656:17:Warning: reference to free variable
    `sniem-last-point-locked'
sniem-operation.el:657:36:Warning: assignment to free variable
    `sniem-last-point'
In end of data:
sniem-operation.el:662:1:Warning: the following functions are not known to be
    defined: sniem-current-mode, sniem-change-mode, sniem-prev-line,
    sniem-end-of-line, sniem-lock/unlock-last-point, sniem-backward-char,
    sniem-end-of-mark, sniem-goto-line, sniem-object-catch--get-second-char,
    sniem-define-motion, &optional, sniem-forward-char, sniem-next-line,
    sniem-motion-hint, sniem-digit-argument-get
```

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20210127.158):
```
32 issues found:
6:0: error: Package-Requires outside the main file have no effect.
8:3: warning: You should include standard keywords: see the variable `finder-known-keywords'.
100:3: error: You should depend on (emacs "24.1") if you need `pcase'.
108:3: error: You should depend on (emacs "24.1") if you need `pcase'.
113:6: error: You should depend on (emacs "24.3") if you need `setq-local'.
123:39: error: You should depend on (emacs "24.3") if you need `setq-local'.
125:0: error: `sniem-up/down-case' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
132:5: error: You should depend on (emacs "24.1") if you need `delete-forward-char'.
138:3: error: You should depend on (emacs "24.1") if you need `delete-forward-char'.
164:3: error: You should depend on (emacs "24.1") if you need `pcase'.
206:3: error: You should depend on (emacs "24.1") if you need `pcase'.
227:3: error: You should depend on (emacs "24.1") if you need `pcase'.
260:15: error: You should depend on (emacs "24.1") if you need `pcase'.
319:13: error: You should depend on (emacs "25.1") if you need `save-mark-and-excursion'.
327:9: error: You should depend on (emacs "24.3") if you need `user-error'.
344:18: error: You should depend on (emacs "25.1") if you need `save-mark-and-excursion'.
348:18: error: You should depend on (emacs "25.1") if you need `save-mark-and-excursion'.
354:13: error: You should depend on (emacs "24.3") if you need `setq-local'.
360:15: error: You should depend on (emacs "24.3") if you need `setq-local'.
362:13: error: You should depend on (emacs "24.3") if you need `setq-local'.
367:9: error: You should depend on (emacs "24.3") if you need `setq-local'.
369:5: error: You should depend on (emacs "24.1") if you need `pcase'.
377:17: error: You should depend on (emacs "24.3") if you need `setq-local'.
389:9: error: You should depend on (emacs "24.3") if you need `user-error'.
390:7: error: You should depend on (emacs "25.1") if you need `save-mark-and-excursion'.
509:3: error: You should depend on (emacs "24.1") if you need `scroll-up-command'.
516:3: error: You should depend on (emacs "24.1") if you need `scroll-down-command'.
554:14: error: You should depend on (emacs "24.1") if you need `pcase'.
557:19: error: You should depend on (emacs "24.3") if you need `user-error'.
572:28: error: You should depend on (emacs "24.3") if you need `setq-local'.
591:28: error: You should depend on (emacs "24.3") if you need `setq-local'.
657:7: error: You should depend on (emacs "24.3") if you need `setq-local'.
```

Other [lints](https://github.com/riscy/melpazoid):
- sniem-operation.el#L121: There is no need to quote lambdas (neither #' nor ')
- sniem-operation.el#L576: Use `search-*`'s NOERROR argument
- sniem-operation.el#L121: Loading a package should rarely add hooks
- sniem-operation.el#L374: Loading a package should rarely add advice

### sniem-var.el ###

byte-compile (using Emacs 27.1):
```
sniem-var.el:257:26:Warning: reference to free variable
    `sniem-object-catch-forward-p'
sniem-var.el:264:17:Warning: reference to free variable
    `awesome-tray-module-alist'
sniem-var.el:264:17:Warning: assignment to free variable
    `awesome-tray-module-alist'
In end of data:
sniem-var.el:269:1:Warning: the function `sniem-current-mode' is not known to
    be defined.
```

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20210127.158):
```
4 issues found:
6:0: error: Package-Requires outside the main file have no effect.
8:3: warning: You should include standard keywords: see the variable `finder-known-keywords'.
194:51: error: This key sequence is reserved (see Key Binding Conventions in the Emacs Lisp manual)
255:3: error: You should depend on (emacs "24.1") if you need `pcase'.
```

### sniem.el ###

byte-compile (using Emacs 27.1):
```
sniem.el:42:1:Error: Symbol's value as variable is void: sniem-beginning-of-line
```

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20210127.158):
```
34 issues found:
1:50: warning: You should depend on (emacs "24.1") if you need lexical-binding.
6:0: error: Expected (package-name "version-num"), but found (emacs).
6:31: warning: Use a non-snapshot version number for dependency on "s" if possible.
6:50: warning: Use a non-snapshot version number for dependency on "dash" if possible.
8:3: warning: You should include standard keywords: see the variable `finder-known-keywords'.
52:0: error: Global minor modes should be autoloaded or, rarely, `:require' their defining file (i.e. ":require 'sniem"), to support the customization variable of the same name.
79:5: error: You should depend on (emacs "24.3") if you need `setq-local'.
87:5: error: You should depend on (emacs "24.3") if you need `setq-local'.
127:14: error: You should depend on (emacs "24.1") if you need `pcase'.
154:3: error: You should depend on (emacs "24.3") if you need `setq-local'.
174:5: error: You should depend on (emacs "24.1") if you need `pcase'.
182:3: error: You should depend on (emacs "24.3") if you need `setq-local'.
182:27: error: You should depend on (emacs "24.1") if you need `pcase'.
224:3: error: You should depend on (emacs "24.1") if you need `pcase'.
282:8: error: You should depend on (emacs "24.3") if you need `user-error'.
290:5: error: You should depend on (emacs "25.1") if you need `prefix-command-preserve-state'.
292:5: error: You should depend on (emacs "24.4") if you need `universal-argument--mode'.
314:19: error: You should depend on (emacs "24.4") if you need `string-empty-p'.
319:14: error: You should depend on (emacs "25.1") if you need `funcall-interactively'.
326:3: error: You should depend on (emacs "24.1") if you need `pcase'.
338:3: error: You should depend on (emacs "24.1") if you need `pcase'.
340:6: error: You should depend on (emacs "24.1") if you need `pcase'.
346:6: error: You should depend on (emacs "24.1") if you need `pcase'.
352:6: error: You should depend on (emacs "24.1") if you need `pcase'.
358:0: error: `sniem-lock/unlock-last-point' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
361:3: error: You should depend on (emacs "24.3") if you need `setq-local'.
369:0: error: `sniem-lock/unlock-last-goto-point' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
372:3: error: You should depend on (emacs "24.3") if you need `setq-local'.
388:11: error: You should depend on (emacs "24.3") if you need `setq-local'.
389:7: error: You should depend on (emacs "24.3") if you need `setq-local'.
399:5: error: You should depend on (emacs "25.1") if you need `save-mark-and-excursion'.
409:51: error: You should depend on (emacs "24.1") if you need `pcase'.
418:5: error: You should depend on (emacs "24.3") if you need `setq-local'.
424:5: error: You should depend on (emacs "25.1") if you need `funcall-interactively'.
```

Other [lints](https://github.com/riscy/melpazoid):
- sniem.el#L130: Optionally use `unless ...` instead of `when (null ...)`
- sniem.el#L441: Loading a package should rarely add advice
- sniem.el#L442: Loading a package should rarely add advice

**Comment by SpringHan** [comment]
I've solve the problems, and the warnings of `sniem-operation.el` caused by the macro I defined, so it's not need to modify.  
Thanks the contribution for me!!!

**Comment by riscy** [comment]
Thank ya.  Enclosing the latest checks below --

### sniem-mark-jump.el ###

byte-compile (using Emacs 27.1):
```
sniem-mark-jump.el:93:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-mark-jump.el:93:22:Warning: reference to free variable
    `sniem-mark-jump-next'
sniem-mark-jump.el:93:54:Warning: reference to free variable `type'
sniem-mark-jump.el:98:28:Warning: assignment to free variable `type'
sniem-mark-jump.el:100:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-mark-jump.el:100:22:Warning: reference to free variable
    `sniem-mark-jump-prev'
In end of data:
sniem-mark-jump.el:187:1:Warning: the following functions are not known to be
    defined: sniem-open-line-previous, sniem-object-catch-lisp-mode-p,
    sniem-change-mode, sniem-define-motion, &optional
```

### sniem-object-catch.el ###

byte-compile (using Emacs 27.1):
```
sniem-object-catch.el:39:1:Warning: defcustom for
    `sniem-object-catch-global-symbol-alist' fails to specify type
sniem-object-catch.el:39:1:Warning: defcustom for
    `sniem-object-catch-global-symbol-alist' fails to specify type
In sniem-object-catch-parent-by-char:
sniem-object-catch.el:177:32:Warning: `message' called with 0 args to fill 1
    format field(s)
In prog3:
sniem-object-catch.el:298:11:Warning: macro `prog3' defined too late
sniem-object-catch.el:307:1:Warning: Lexical argument shadows the dynamic
    variable mode-name
In end of data:
sniem-object-catch.el:336:1:Warning: the function `sniem-normal-set-key' is
    not known to be defined.
```

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20210127.158):
```
3 issues found:
6:0: error: Package-Requires outside the main file have no effect.
69:0: error: "global-sniem-object-catch-status" doesn't start with package's prefix "sniem".
298:0: error: "prog3" doesn't start with package's prefix "sniem".
```
(The "Package-Requires outside the main file have no effect" is due to my setting `package-lint-main-file` before the run; you can do similarly, and something to be aware of.)

### sniem-operation.el ###

byte-compile (using Emacs 27.1):
```
sniem-operation.el:93:10:Warning: `beginning-of-buffer' is for interactive use
    only; use `(goto-char (point-min))' instead.
In sniem-up/down-case:
sniem-operation.el:134:13:Warning: `delete-forward-char' is for interactive
    use only; use `delete-char' instead.
In sniem-replace-char:
sniem-operation.el:140:4:Warning: `delete-forward-char' is for interactive use
    only; use `delete-char' instead.
In sniem-macro:
sniem-operation.el:375:33:Warning: name-last-kbd-macro called with 0
    arguments, but requires 1
sniem-operation.el:382:1:Warning: Unused lexical variable `second-char'
In sniem-prev-line:
sniem-operation.el:458:15:Warning: `previous-line' is for interactive use
    only; use `forward-line' with negative argument instead.
In sniem-next-line:
sniem-operation.el:471:15:Warning: `next-line' is for interactive use only;
    use `forward-line' instead.
sniem-operation.el:555:1:Warning: Unused lexical variable `point'
sniem-operation.el:574:1:Warning: Unused lexical variable `point'
In end of data:
sniem-operation.el:653:1:Warning: the following functions are not known to be
    defined: sniem-current-mode, sniem-change-mode,
    sniem-lock-unlock-last-point, sniem-object-catch--get-second-char,
    sniem-motion-hint, sniem-digit-argument-get
```

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20210127.158):
```
2 issues found:
6:0: error: Package-Requires outside the main file have no effect.
128:0: error: `sniem-up/down-case' contains a non-standard separator `/', use hyphens instead (see Elisp Coding Conventions).
```

### sniem-var.el ###

byte-compile (using Emacs 27.1):
- No issues!

checkdoc (using version 0.6.2):
- No issues!

package-lint (using version 20210127.158):
```
2 issues found:
6:0: error: Package-Requires outside the main file have no effect.
194:51: error: This key sequence is reserved (see Key Binding Conventions in the Emacs Lisp manual)
```

### sniem.el ###

byte-compile (using Emacs 27.1):
```
sniem.el:437:1:Error: Symbol's function definition is void: sniem-normal-set-key
```

**Comment by SpringHan** [comment]
First, sorry, we are in the different country and time zone, so we can't communicate at the same time.  
I've solved some of the issuses, and the contents below are the explain for some warnings and error.

## sniem-mark-jump

byte-compile:

```
sniem-mark-jump.el:93:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-mark-jump.el:93:22:Warning: reference to free variable
    `sniem-mark-jump-next'
sniem-mark-jump.el:93:54:Warning: reference to free variable `type'
sniem-mark-jump.el:98:28:Warning: assignment to free variable `type'
sniem-mark-jump.el:100:1:Warning: misplaced interactive spec: `(interactive
    "P")'
sniem-mark-jump.el:100:22:Warning: reference to free variable
    `sniem-mark-jump-prev'
```
These warnings caused by the macro I defined, and in the macro, there has the (interactive) spec. So they don't have some problems.

## sniem.el

byte-compile:

```
sniem.el:437:1:Error: Symbol's function definition is void: sniem-normal-set-key
```
This error I haven't find the origin. And in all the project, this function only be used for three times in the same file.(In this file, I defined the function.)

**Comment by SpringHan** [comment]
I created a new branch to add some new feature, so I added the `branch` argument.

**Comment by riscy** [comment]
Thanks, I'm seeing the improvements.

MELPA will pull your repository's default branch, so most likely the branch doesn't need to be specified in your recipe.  In fact, we prefer that you don't specify a branch, since this will make it easy for you to rename your default branch in the future without a followup PR.

> These warnings caused by the macro I defined, and in the macro, there has the (interactive) spec. So they don't have some problems.

Trust the compiler on this one -- you can expand the macro using `macroexpand` which should give you a clearer picture of why the warnings are being raised.

> When I byte-compile this file, it only output ...

Most likely you're not seeing all of the compilation warnings because you're compiling your code in an instance of Emacs that has already loaded the code, or some past version of it at least.  Using `flycheck` helps with this, or you can restart emacs using `emacs -Q` and try compiling from there.

> This error I haven't find the origin. And in all the project, this function only be used for three times in the same file.

Taking a quick look, this is because `sniem-set-keyboard-layout` contains the `pcase` macro, and the compiler tries to expand/evaluate macros before it looks at the defuns.   I think the simplest way forward here would just be to use a `cond` form instead of `pcase`.

File: recipes/sniem [hunk 2]
```diff
@@ -0,0 +1 @@
+(sniem :repo "SpringHan/sniem" :fetcher github :branch "master")
```
Comment (SpringHan) [inline recipes/sniem:1]:
```suggestion
(sniem :repo "SpringHan/sniem" :fetcher github)
```

**Comment by SpringHan** [comment]
Thanks, the way works well. Now I've edited the package and there's no error or warning when compiling the project.  
Thanks for your contribution!

**Comment by riscy** [comment]
Here's the latest --

### sniem-common.el ###

byte-compile (using Emacs 27.1):
```
sniem-common.el:30:10:Warning: reference to free variable `sniem-normal-mode'
sniem-common.el:31:10:Warning: reference to free variable `sniem-insert-mode'
sniem-common.el:32:10:Warning: reference to free variable `sniem-motion-mode'
In sniem-object-catch--get-second-char:
sniem-common.el:58:24:Warning: reference to free variable
    `sniem-object-catch-global-symbol-alist'
In end of data:
sniem-common.el:140:1:Warning: the following functions are not known to be
    defined: sniem-normal-mode, sniem-insert-mode, sniem-motion-mode,
    sniem-cursor-change, sniem-show-last-point,
    sniem-digit-argument-read-char, sniem-digit-argument-fn-get
```

### sniem-mark-jump.el ###

byte-compile (using Emacs 27.1):
```
sniem-mark-jump.el:182:1:Warning: the function
    `sniem-object-catch-lisp-mode-p' is not known to be defined.
```

### sniem-object-catch.el ###
package-lint (using version 20210127.158):
```
1 issue found:
246:0: error: "prog3" doesn't start with package's prefix "sniem".
```


### sniem.el ###

byte-compile (using Emacs 27.1):
```
sniem.el:278:5:Warning: reference to free variable `_'
sniem.el:390:17:Warning: reference to free variable
    `awesome-tray-module-alist'
sniem.el:390:17:Warning: assignment to free variable
    `awesome-tray-module-alist'
```

I'll also mention that loading the package itself should not automatically add hooks and advice.  See some of my previous comments for where these are coming into play.  Toggling these features on and off should probably be done using a global minor mode -- let me know if that needs any clarification.

**Comment by SpringHan** [comment]
I've solved some of the problems.  
But to some warnings, I'll say my opinion.

## sniem-common.el
```
sniem-common.el:30:10:Warning: reference to free variable `sniem-normal-mode'
sniem-common.el:31:10:Warning: reference to free variable `sniem-insert-mode'
sniem-common.el:32:10:Warning: reference to free variable `sniem-motion-mode'
In sniem-object-catch--get-second-char:
sniem-common.el:58:24:Warning: reference to free variable
    `sniem-object-catch-global-symbol-alist'
In end of data:
sniem-common.el:140:1:Warning: the following functions are not known to be
    defined: sniem-normal-mode, sniem-insert-mode, sniem-motion-mode,
    sniem-cursor-change, sniem-show-last-point,
    sniem-digit-argument-read-char, sniem-digit-argument-fn-get
```
To solve these warnings is a little difficult, and not I don't have much time. (The reason is at the bottom)

## sniem.el
```
sniem.el:390:17:Warning: reference to free variable
    `awesome-tray-module-alist'
sniem.el:390:17:Warning: assignment to free variable
    `awesome-tray-module-alist'.
```
In my project, I provide the support for the package named `awesome-tray`, so if I want to save the support, I can't delete these codes. And I've also added the check with `featurep`.

## The advice
I use the advice in package to provide support for `dired-mode` and give user some functions.

If you mind the warnings or others, you can close this pull request. Now I'll go to school so that I can only use computer on weekends.  
Last, thanks for your contribution again.

**Comment by riscy** [comment]
> To solve these warnings is a little difficult, and not I don't have much time.... If you mind the warnings or others, you can close this pull request. Now I'll go to school so that I can only use computer on weekends.

Agreed -- it can be lots of work to maintain packages.  By the look of your GitHub stars you do have a few interested users, so you could possibly create an issue that asks for help.

At your suggestion I'll close the PR for now, but you can always feel free to reopen or create a new PR if you want another look.  The tool I'm using to run the above checks is [here](https://github.com/riscy/melpazoid) and may be helpful.

**Comment by SpringHan** [comment]
Thesedays, I have time to manage my package, so open the pull requests please.
Today and tommorrw, I'll try to solve the problems in the package.

**Comment by SpringHan** [comment]
The lastest commit has solved the prolems.

> ## sniem.el
> ```
> sniem.el:390:17:Warning: reference to free variable
>     `awesome-tray-module-alist'
> sniem.el:390:17:Warning: assignment to free variable
>     `awesome-tray-module-alist'.
> ```
> 
> In my project, I provide the support for the package named `awesome-tray`, so if I want to save the support, I can't delete these codes. And I've also added the check with `featurep`.

There's only this warning.

**Comment by SpringHan** [comment]
@riscy 

It seems that only the admin of the repository can reopen the PR as this PR was closed by you.

**Comment by riscy** [comment]
Yes, I'll gladly open this PR - I unchecked all the checkboxes in the PR description, so take a quick walk through the checklist and update those when you have a chance.  Thanks.

**Comment by SpringHan** [comment]
Thanks, and I found the check has some error.  
It says that the `cl-lib` package is not available. But the `cl-lib` package is a built-in package in emacs, or the package was to be a built-in package after emacs25. (Because in the check program, I found the emacs it installed is emacs25.)  
I found a PR used `cl-lib` package also has this error.

Emmm, I don't know the reason.

**Comment by riscy** [comment]
> Thanks, and I found the check has some error.

Sorry about that - there was a change to a tool used in our CI that will cause this branch to be unable to build until you either merge the master branch or rebase on top of it.

**Comment by SpringHan** [comment]
I'll create a new PR to solve it.

**Pull request closed without merging**